src/dill/gauge-proxy-v2.sol

```diff
@@ -0,0 +1,2161 @@
+        private userRewardPerTokenPaid;
+    mapping(address => mapping(uint256 => uint256)) public _rewards;
+    mapping(address => bool) public authorisedAddress;
+    uint256[] private rewardPerTokenStored;
+    uint256 public multiplierDecayPerSecond = uint256(48e9);
+    mapping(address => mapping(uint256 => uint256)) private _lastUsedMultiplier;
+    mapping(address => uint256) private _lastRewardClaimTime; // staker addr -> timestamp
+
+    // Balance tracking
+    uint256 public derivedSupply;
+    mapping(address => uint256) public derivedBalances;
+    mapping(address => uint256) private _base;
+
+    // Stake tracking
+    mapping(address => LockedStake[]) private _lockedStakes;
+
+    // Administrative booleans
+    bool public stakesUnlocked; // Release locked stakes in case of emergency
+    mapping(address => bool) public stakesUnlockedForAccount; // Release locked stakes of an account in case of emergency
+
+    /* ========== STRUCTS ========== */
+
+    struct LockedStake {
+        uint256 start_timestamp;
+        uint256 liquidity;
+        uint256 ending_timestamp;
+        uint256 lock_multiplier;
+        bool isPermanentlyLocked;
+    }
+
+    /* ========== MODIFIERS ========== */
+
+    modifier onlyDistribution() {
+        require(
+            msg.sender == DISTRIBUTION,
+            "Caller is not RewardsDistribution contract"
+        );
+        _;
+    }
+
+    modifier onlyGov() {
+        require(
+            msg.sender == governance,
+            "Operation allowed by only governance"
+        );
+        _;
+    }
+
+    modifier lockable(uint256 secs) {
+        require(secs >= lockTimeMin, "Minimum stake time not met");
+        require(
+            secs <= lockTimeForMaxMultiplier,
+            "Trying to lock for too long"
+        );
+        _;
+    }
+
+    modifier updateReward(address account, bool isClaimReward) {
+        rewardPerTokenStored = rewardPerToken();
+        lastUpdateTime = lastTimeRewardApplicable();
+
+        if (account != address(0)) {
+            uint256[] memory earnedArr = earned(account);
+            for (uint256 i = 0; i < rewardPerTokenStored.length; i++) {
+                _rewards[account][i] = earnedArr[i];
+                userRewardPerTokenPaid[account][i] = rewardPerTokenStored[i];
+            }
+        }
+        _;
+        if (account != address(0)) {
+            kick(account);
+            if (isClaimReward) {
+                _lastRewardClaimTime[account] = block.timestamp;
+            }
+        }
+    }
+
+    modifier onlyJarAndAuthorised() {
+        require(
+            msg.sender == address(jar) || authorisedAddress[msg.sender],
+            "Sender not an authorized jar or address."
+        );
+        _;
+    }
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _jar,
+        address _governance,
+        address _distribution,
+        string[] memory _rewardSymbols,
```

Comment:
I think we shouldn't be trying to store arbitrary `rewardSymbols` in our contracts - contract addresses should be the first class citizens (in events and whatever else). If needed, the cononical symbols can be obtained using `IERC20(token).symbol()`

Reply:
Fixed this.